### PR TITLE
feat: add ListTopics use case

### LINKED ISSUE #15: Create an use case to list all topics

### PR DESCRIPTION
**Pull request by sptGabriel**
this pr adds a new feature to the use case, to list all topics from use case storage dependency.

closes #15

### CODE REVIEW

File: repositories/memory/topics/list_topics.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package topics
+
+import (
+	"context"
+
+	"github.com/matheusmosca/walrus/domain/entities"
+)
+
+func (r *repository) ListTopics(_ context.Context) ([]entities.Topic, error) {
+	topics := make([]entities.Topic, 0, len(r.storage))
```
Comment (matheusmosca) [inline repositories/memory/topics/list_topics.go:10]: If the storage length is 0, then an error should be returned. You can create an error in the domain layer like ErrNoTopicsFound and return it here. You can also use this error on unit tests asserting it with ErrorIs

File: domain/usecases/list_topics_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			useCase := New(tt.fields.storage)
+			topics, err := useCase.ListTopics(tt.args.ctx)
+			if tt.wantErr {
+				assert.Error(t, err)
```
Comment (matheusmosca) [inline domain/usecases/list_topics_test.go]:
```suggestion
				assert.Error(t, err)
				assert.Empty(t, topics)
```

File: domain/usecases/list_topics_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		},
+		{
+			name: "empty storage topic should success",
+			args: args{
+				ctx: context.Background(),
+			},
+			fields: fields{
+				storage: &RepositoryMock{
+					ListTopicsFunc: func(ctx context.Context) ([]entities.Topic, error) {
+						return []entities.Topic{}, nil
```
Comment (matheusmosca) [inline domain/usecases/list_topics_test.go]:
```suggestion
						return nil, nil
```

File: repositories/memory/topics/list_topics.go (hunk 1 above)
Comment (sptGabriel) [reply repositories/memory/topics/list_topics.go:10]: do you think an empty list should return an error?

Comment (sptGabriel) [reply repositories/memory/topics/list_topics.go:10]: I thought about error reasons like (some problem with SGBDs), but if you think it's better I can return the error to empty list

File: domain/usecases/list_topics_test.go (hunk 2 above)
Comment (sptGabriel) [reply domain/usecases/list_topics_test.go]: done

File: domain/usecases/list_topics_test.go (hunk 3 above)
Comment (sptGabriel) [reply domain/usecases/list_topics_test.go]: done

File: domain/usecases/list_topics_test.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		storage Repository
+	}
+
+	tests := []struct {
+		name      string
+		args      args
+		fields    fields
+		beforeRun func(topic entities.Topic) chan vos.Message
+		want      []vos.TopicName
+		wantErr   bool
```
Comment (matheusmosca) [inline domain/usecases/list_topics_test.go]: could you change this to be an error type? and check this error with ErrorIs? The second test which the repository returns `errors.New("wrong")` could be deleted.

Comment (sptGabriel) [reply domain/usecases/list_topics_test.go]: done

File: domain/entities/errors.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package entities
 import "errors"
 
 var (
+	ErrNoTopicsFound      = errors.New("no topic was found")
```
Comment (MelloTonio) [inline domain/entities/errors.go]:
```suggestion
	ErrTopicsNotFound = errors.New("no topic was found")
```

**Review by MelloTonio — APPROVED** [review]

File: domain/usecases/list_topics_test.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+			useCase := New(tt.fields.storage)
+			topics, err := useCase.ListTopics(tt.args.ctx)
+			if tt.wantErr != nil {
+				assert.ErrorIs(t, err, tt.wantErr)
+				assert.Empty(t, topics)
+				return
+			}
+
+			require.NoError(t, err)
```
Comment (matheusmosca) [inline domain/usecases/list_topics_test.go]:
```suggestion
			assert.ErrorIs(t, err, tt.wantErr)
```

File: domain/entities/errors.go (hunk 5 above)
Comment (sptGabriel) [reply domain/entities/errors.go]: @MelloTonio thanks.

File: domain/usecases/list_topics_test.go (hunk 6 above)
Comment (sptGabriel) [reply domain/usecases/list_topics_test.go]: @matheusmosca done

**Review by matheusmosca — APPROVED** [review]

**Pull request merged**